_posts/2017-09-24-Unit-testing-in-angular.md

```diff
@@ -0,0 +1,504 @@
+The routes that are relevant can be defined in the `RouterTestingModule`:
+    
+{% highlight coffeescript %}  
+    imports: [RouterTestingModule.withRoutes([/*List mock routes here*/])]
+{% endhighlight %}  
+
+If you don't feel like doing all that, you can also tell Angular to skip elements it doesn't know by adding `NO_ERRORS_SCHEMA` to the TestBed configuration:
+
+{% highlight coffeescript %}  
+    TestBed.configureTestingModule({
+        declarations: [ AppComponent ],
+        schemas: [ NO_ERRORS_SCHEMA ]
+    })
+{% endhighlight %}  
+
+To learn more about writing tests using Angular TestBed, I recommend reading this guide: [https://angular.io/guide/testing](https://angular.io/guide/testing).
+
+## 2. Calling the constructor
+
+A much simpler way to do unit testing is to simply call the constructor of the class you want to test.
+You should get an instance of each class that's needed in the constructor.
+Of course we want to mock these classes and as we saw in the Angular TestBed section, the Jasmine spy objects are the way to go.
+
+{% highlight coffeescript %}  
+    describe('AppComponent', () => {        
+        let component: AppComponent;
+
+        const mockCarBrandService = jasmine.createSpyObj('carBrandService', ['findAll']);
+        mockCarBrandService.findAll.and.returnValue(Observable.of([
+            { name: 'Mazda', country: 'Japan' },
+            { name: 'BMW', country: 'Germany' }
+        ]);
+
+        beforeEach(() => {
+            component = new AppComponent(mockCarBrandService);
+        });
+        ...
+    });
+{% endhighlight %}  
+
+Without the TestBed, you don’t have access to the view. 
+However, your tests will run much faster as there are less things to load. 
+When using TestBed, you’ll probably be including lots of dependencies just to make it work, giving you less control. 
+Something you do not want in unit testing. 
+Another difference with TestBed is that you have to call the lifecycle events yourself, again giving you more control over the code you’re testing.
+
+{% highlight coffeescript %}  
+    it('should find the car brand', () => {
+        component.ngOnInit();
+        const carBrand = component.getFavoriteCarBrand();
+        expect(carBrand.name).toEqual('Mazda');
+    }); 
+{% endhighlight %}  
+
+These kind of tests are much easier to write, give you more control, run faster, allow you to completely isolate the class and take away the hassle of getting the TestBed setup right. 
+Unless of course, you really want to test the view as well. 
+However, in my opinion, that’s not the point of unit testing. 
+You have protactor (end to end / integration) tests to test the view.
+
+## Async, fakeAsync, tick
+Angular is full of Observables and writing tests for them is a little trickier. 
+You might also be using `setTimeout` and `setInterval`. 
+To cope with all that, there's the `async` and `fakeAsync` function. 
```

Comment:
not clear where async, fakeasync, tick. are coming from. make a note that they come with the angular library (not jasmine) and are related with zone.js